services/galley/src/Galley/Data.hs

```diff
@@ -551,6 +552,30 @@ conversationIdsFrom usr start (fromRange -> max) =
   where
     strip p = p {result = take (fromIntegral max) (result p)}
 
+-- | When the 'start' parameter is set to 'Nothing', reads first 'max' records
+-- from 'user_remote_converstaions' table.
+--
+-- Otherwise, reads 'max' records starting from the 'start' parameter. Doing
+-- this is unfortunately not trivial, so this function first gets all the
+-- conversations which match the domain and then if there is still space, gets
+-- the conversations which have domain > domain of start.
```

Comment:
Not exactly this, but there is a `token` function. But its behaviour depends on some Cassandra settings, I am going to try to use it and if it works, maybe explore how Cassandra will behave under different settings.

Reply:
From my reading so far `token` can only be used on "partitiion keys". Partition keys are keys which are listed in the first column of the primary key. Right now our primary key is `(user, remote_conv_domain, remote_conv_id)`, so our primary key is just `user`. If we want this to work, I guess we'll have to change this to something more complex like `((user, remote_conv_domain, remote_conv_id), user)`. And then set the cluster ordering by `user`. This may allow us to write queries like `WHERE token(user, remote_conv_domain, remote_conv_id) > (?, ?, ?) AND user = ?`. Not sure what other problems this might cause, I am already thinking this might cause data for one user to be scattered across different nodes, which might not be ideal. I am gonna read more.

Reply:
Actually, the partition key doesn't need to be unique, so we can even get rid of `user` from it, making our primary key to be `((remote_conv_domain, remote_conv_id), user)`

Reply:
But why not use `pagingState` instead and let the driver handle it? Or am I missing something?

Reply:
> But why not use pagingState instead and let the driver handle it? Or am I missing something?

I will try this next and see if it works.

Reply:
This actually works :tada: